sourcegraph/sourcegraph
# insights: prune stream insights snapshots

closes #32969.

We prune the snapshots table when adding new data series to search insights (not capture group insights). A snapshot is the latest retrieved data series for an insight, and an insight is resolved from the union between `series_points` and `series_points_snapshots`.

Search stream insights are a backend-enabled only functionality for which we forgot to add this logic to prune the snapshots table.

This also moves some helpers around in `work_handler` as the file was hard to read in a logical way.

## Test plan

The behaviour can be reproduced as follows, keeping in mind that search stream insights can only be enabled by going into the database at the moment. 

1. Create a search insight
2. Set it to search stream manually
    * To enable streaming for an insight, its capture group method should be set to `search-stream` and the`backfill_queued_at` timestamp should be nulled. [More specific instructions](https://sourcegraph.slite.com/app/docs/vkc5e0tHs-s5jB)
3. Run a few snapshots by setting the next snapshot after timestamp to `NOW()` and re-running the query runner.
4. Observe database and insights state before, and after the change.

Screenshots:
Multiple values for the same series, at different times:
<img width="558" alt="Step 4 - observe there are duplicate snapshots" src="https://user-images.githubusercontent.com/29406849/162725898-ec3f9855-d7f8-42cf-a885-95091261b29f.png">
After fix, values get overwritten:
<img width="1010" alt="Screenshot 2022-04-07 at 14 49 20" src="https://user-images.githubusercontent.com/29406849/162726220-7716eb1e-77de-41d6-a5bc-0e4ac10323bf.png">
<img width="1112" alt="Screenshot 2022-04-07 at 14 48 20" src="https://user-images.githubusercontent.com/29406849/162726208-47f6437b-2bbe-4ac4-aae0-5f61e380f6df.png">
